contracts/bridge/RateLimiter.sol

```diff
@@ -0,0 +1,307 @@
+        }
+        allowance.resetTimeMin = resetTimeMin;
+        allowance.amount = allowanceAmount;
+        updateAllowance(token, allowance);
+        emit SetAllowance(token, allowanceAmount, resetTimeMin);
+    }
+
+    function updateAllowance(address token, Allowance memory allowance)
+        private
+    {
+        allowances[token] = allowance;
+    }
+
+    function getAllowance(address token)
+        private
+        view
+        returns (Allowance memory allowance)
+    {
+        allowance = allowances[token];
+        // solium-disable-next-line security/no-block-members
+        uint32 currentMin = uint32(block.timestamp / 60);
+        // Check if we should reset the time. We do this on load to minimize storage read/ writes
+        if (
+            allowance.resetTimeMin > 0 &&
+            allowance.lastResetMin <= currentMin - allowance.resetTimeMin
+        ) {
+            allowance.spent = 0;
+            // Resets happen in regular intervals and `lastResetMin` should be aligned to that
+            allowance.lastResetMin =
+                currentMin -
+                ((currentMin - allowance.lastResetMin) %
+                    allowance.resetTimeMin);
+        }
+        return allowance;
+    }
+
+    /**
+     * @notice Checks the allowance for a given token. If the new amount exceeds the allowance, it is not updated and false is returned
+     * otherwise true is returned and the transaction can proceed
+     * @param amount to transfer
+     **/
+    function checkAndUpdateAllowance(address token, uint256 amount)
+        external
+        nonReentrant
+        onlyRole(BRIDGE_ROLE)
+        returns (bool)
+    {
+        Allowance memory allowance = getAllowance(token);
+
+        // Update state
+        // @dev reverts if amount > (2^96 - 1)
+        uint96 newSpent = allowance.spent + uint96(amount);
+
+        // do not proceed. Store the transaction for later
+        if (newSpent > allowance.amount) {
+            return false;
+        }
+
+        allowance.spent = newSpent;
+        updateAllowance(token, allowance);
+
+        return true;
+    }
+
+    function addToRetryQueue(bytes32 kappa, bytes memory toRetry)
+        external
+        onlyRole(BRIDGE_ROLE)
+    {
+        rateLimitedQueue.add(kappa, toRetry);
+    }
+
+    function retryByKappa(bytes32 kappa) external {
+        (bytes memory toRetry, uint32 storedAtMin) = rateLimitedQueue.get(
+            kappa
+        );
+        if (toRetry.length > 0) {
+            if (!hasRole(LIMITER_ROLE, msg.sender)) {
+                // Permissionless retry is only available once timeout is finished
+                uint32 currentMin = uint32(block.timestamp / 60);
+                require(
+                    currentMin >= storedAtMin + retryTimeout,
+                    "Retry timeout not finished"
+                );
+            }
+            _retry(kappa, toRetry);
+            rateLimitedQueue.deleteKey(kappa);
```

Comment:
Maybe re-order 208 & 209 to modify state of rateLimitedQueue first, then attempt call to bridge.

Reply:
Resolved in [bdbc189](https://github.com/synapsecns/synapse-contracts/pull/116/commits/bdbc1891fe81f9a213b2b3158bd612a76f310387)